### PR TITLE
[wasm][bcl] Use zlib for DeflateStream on WASM profile

### PR DESCRIPTION
**Pull request by kjpou1**
- Add `-s USE-ZLIB=1` flag for the emscripten build of `mono.js`
- Repurpose `zlib-helper.c`, from `mono/support`, and include it in the `mono` wasm build.
- Modify `packager.exe` to include building zlib support for aot.
- Modify `wasm.mk` to copy the zlib headers needed to build in the zlib support so they are included in the packaging of the runtime.
- Add build targets to test zlib support for interpreted and aot build.  These are helper targets for testing.
- Add tests for System.IO.Compression through browser. 
- Add new option to packager to enable zlib support
   -   --zlib (enable the use of zlib for System.IO.Compression support)
           type: bool  default: false
- Debug build sizes without `-s USE_ZLIB=1`

| size | date | name |
|---|---|---|
| 567306 | May 27 08:18 | mono.js |
| 2202385 | May 27 08:18 | mono.wasm |
| 444381 | May 27 08:18 | mono.wasm.map |
| 38168662 | May 27 08:18 | mono.wast |

- Debug build sizes with `-s USE_ZLIB=1`

| size | date | name |
|---|---|---|
| 567306 | May 27 12:48 | mono.js |
| 2260089 | May 27 12:48 | mono.wasm |
| 445324 | May 27 12:48 | mono.wasm.map |
| 39029480 | May 27 12:48 | mono.wast |

- Debug build size difference:

` 39029480 - 38168662 = 860818`

- Release build sizes without `-s USE_ZLIB=1`

| size | date | name |
|---|---|---|
| 240418 | May 27 08:19 | mono.js |
| 1759533 | May 27 08:19 | mono.wasm |

- Release build sizes with `-s USE_ZLIB=1`

| size | date | name |
|---|---|---|
| 240418 | May 27 12:48 | mono.js |
| 1815925 | May 27 12:48 | mono.wasm |

- Release build size difference:

`1815925 - 1759533 =  56392`

Closes issue: https://github.com/mono/mono/issues/8216

### CODE REVIEW

**Comment by kjpou1** [comment]
Right now receiving the following error on the delegate callback:

```

JS exception: RuntimeError: indirect call signature mismatch
@/Users/Jimmy/websharp/projects/mono/sdks/wasm/bin/zlib-sample/mono.js line 1 > WebAssembly.instantiate:wasm-function[2362]:0x86957
@/Users/Jimmy/websharp/projects/mono/sdks/wasm/bin/zlib-sample/mono.js line 1 > WebAssembly.instantiate:wasm-function[1612]:0x5c48d
@/Users/Jimmy/websharp/projects/mono/sdks/wasm/bin/zlib-sample/mono.js line 1 > WebAssembly.instantiate:wasm-function[4506]:0xf821b
@/Users/Jimmy/websharp/projects/mono/sdks/wasm/bin/zlib-sample/mono.js line 1 > WebAssembly.instantiate:wasm-function[6068]:0x134610
@/Users/Jimmy/websharp/projects/mono/sdks/wasm/bin/zlib-sample/mono.js line 1 > WebAssembly.instantiate:wasm-function[1878]:0x688e3
@/Users/Jimmy/websharp/projects/mono/sdks/wasm/bin/zlib-sample/mono.js line 1 > WebAssembly.instantiate:wasm-function[625]:0x1b1be
@/Users/Jimmy/websharp/projects/mono/sdks/wasm/bin/zlib-sample/mono.js line 1 > WebAssembly.instantiate:wasm-function[625]:0x1aceb
@/Users/Jimmy/websharp/projects/mono/sdks/wasm/bin/zlib-sample/mono.js line 1 > WebAssembly.instantiate:wasm-function[625]:0x1ae9a
@/Users/Jimmy/websharp/projects/mono/sdks/wasm/bin/zlib-sample/mono.js line 1 > WebAssembly.instantiate:wasm-function[625]:0x1aceb
@/Users/Jimmy/websharp/projects/mono/sdks/wasm/bin/zlib-sample/mono.js line 1 > WebAssembly.instantiate:wasm-function[625]:0x1aceb
@/Users/Jimmy/websharp/projects/mono/sdks/wasm/bin/zlib-sample/mono.js line 1 > WebAssembly.instantiate:wasm-function[625]:0x1ae9a
@/Users/Jimmy/websharp/projects/mono/sdks/wasm/bin/zlib-sample/mono.js line 1 > WebAssembly.instantiate:wasm-function[625]:0x1aceb
@/Users/Jimmy/websharp/projects/mono/sdks/wasm/bin/zlib-sample/mono.js line 1 > WebAssembly.instantiate:wasm-function[625]:0x1aceb
@/Users/Jimmy/websharp/projects/mono/sdks/wasm/bin/zlib-sample/mono.js line 1 > WebAssembly.instantiate:wasm-function[625]:0x1ae9a
@/Users/Jimmy/websharp/projects/mono/sdks/wasm/bin/zlib-sample/mono.js line 1 > WebAssembly.instantiate:wasm-function[625]:0x1ae9a
@/Users/Jimmy/websharp/projects/mono/sdks/wasm/bin/zlib-sample/mono.js line 1 > WebAssembly.instantiate:wasm-function[625]:0x1aceb
@/Users/Jimmy/websharp/projects/mono/sdks/wasm/bin/zlib-sample/mono.js line 1 > WebAssembly.instantiate:wasm-function[625]:0x1ae9a
@/Users/Jimmy/websharp/projects/mono/sdks/wasm/bin/zlib-sample/mono.js line 1 > WebAssembly.instantiate:wasm-function[625]:0x1ae9a
@/Users/Jimmy/websharp/projects/mono/sdks/wasm/bin/zlib-sample/mono.js line 1 > WebAssembly.instantiate:wasm-function[625]:0x1aceb
@/Users/Jimmy/websharp/projects/mono/sdks/wasm/bin/zlib-sample/mono.js line 1 > WebAssembly.instantiate:wasm-function[625]:0x1ae9a
@/Users/Jimmy/websharp/projects/mono/sdks/wasm/bin/zlib-sample/mono.js line 1 > WebAssembly.instantiate:wasm-function[625]:0x1ae9a
@/Users/Jimmy/websharp/projects/mono/sdks/wasm/bin/zlib-sample/mono.js line 1 > WebAssembly.instantiate:wasm-function[625]:0x1aceb
@/Users/Jimmy/websharp/projects/mono/sdks/wasm/bin/zlib-sample/mono.js line 1 > WebAssembly.instantiate:wasm-function[625]:0x1ae9a
@/Users/Jimmy/websharp/projects/mono/sdks/wasm/bin/zlib-sample/mono.js line 1 > WebAssembly.instantiate:wasm-function[625]:0x1ae9a
@/Users/Jimmy/websharp/projects/mono/sdks/wasm/bin/zlib-sample/mono.js line 1 > WebAssembly.instantiate:wasm-function[625]:0x1ae9a
@/Users/Jimmy/websharp/projects/mono/sdks/wasm/bin/zlib-sample/mono.js line 1 > WebAssembly.instantiate:wasm-function[625]:0x1ae9a
@/Users/Jimmy/websharp/projects/mono/sdks/wasm/bin/zlib-sample/mono.js line 1 > WebAssembly.instantiate:wasm-function[625]:0x1aceb
@/Users/Jimmy/websharp/projects/mono/sdks/wasm/bin/zlib-sample/mono.js line 1 > WebAssembly.instantiate:wasm-function[3451]:0xcaf7f
@/Users/Jimmy/websharp/projects/mono/sdks/wasm/bin/zlib-sample/mono.js line 1 > WebAssembly.instantiate:wasm-function[3657]:0xd2b24
@/Users/Jimmy/websharp/projects/mono/sdks/wasm/bin/zlib-sample/mono.js line 1 > WebAssembly.instantiate:wasm-function[1704]:0x6043c
@/Users/Jimmy/websharp/projects/mono/sdks/wasm/bin/zlib-sample/mono.js line 1 > WebAssembly.instantiate:wasm-function[471]:0xf7ed
@/Users/Jimmy/websharp/projects/mono/sdks/wasm/bin/zlib-sample/mono.js line 1 > WebAssembly.instantiate:wasm-function[4823]:0x104cd3
@/Users/Jimmy/websharp/projects/mono/sdks/wasm/bin/zlib-sample/mono.js line 1 > WebAssembly.instantiate:wasm-function[5946]:0x12eb3e
Module._mono_wasm_invoke_method@/Users/Jimmy/websharp/projects/mono/sdks/wasm/bin/zlib-sample/mono.js:1:215966
init@runtime.js:223:15
onRuntimeInitialized/<@runtime.js:119:9
mono_load_runtime_and_bcl/</<@/Users/Jimmy/websharp/projects/mono/sdks/wasm/bin/zlib-sample/mono.js:1:141400

make: *** [run-zlib-test] Error 1


```

File: sdks/wasm/zlib-helper.c [hunk 1]
```diff
@@ -0,0 +1,287 @@
...
+MONO_API gint CloseZStream (ZStream *zstream);
+MONO_API gint Flush (ZStream *stream);
+MONO_API gint ReadZStream (ZStream *stream, guchar *buffer, gint length);
+MONO_API gint WriteZStream (ZStream *stream, guchar *buffer, gint length);
+static gint flush_internal (ZStream *stream, gboolean is_final);
+
+void zlibhelper_initialize_internals ()
+{
+	mono_add_internal_call ("System.IO.Compression.DeflateStreamNative::CreateZStream", CreateZStream);
+	mono_add_internal_call ("System.IO.Compression.DeflateStreamNative::CloseZStream", CloseZStream);
```
Comment (vargaz) [inline sdks/wasm/zlib-helper.c]: Wouldn't be possible to implement these as normal c# icalls ? Then the linker can link them out.

Comment (kjpou1) [reply sdks/wasm/zlib-helper.c]: See next commit.

File: mono/metadata/icall-decl.h [hunk 2]
```diff
@@ -198,7 +198,7 @@ ICALL_EXPORT void ves_icall_System_Runtime_RuntimeImports_ZeroMemory (guint8*, s
 ICALL_EXPORT void ves_icall_System_Runtime_RuntimeImports_ZeroMemory (guint8*, guint);
 #endif
 ICALL_EXPORT void ves_icall_System_Runtime_RuntimeImports_ecvt_s(char*, size_t, double, int, int*, int*);
-#if defined(ENABLE_MONODROID) || defined(ENABLE_MONOTOUCH)
+#if defined(ENABLE_MONODROID) || defined(ENABLE_MONOTOUCH) || defined(HOST_WASM)
 ICALL_EXPORT gpointer ves_icall_System_IO_Compression_DeflateStreamNative_CreateZStream (gint32 compress, MonoBoolean gzip, gpointer feeder, gpointer data);
```
Comment (vargaz) [inline mono/metadata/icall-decl.h]: This should be TARGET_WASM.

File: mono/metadata/icall-def.h [hunk 3]
```diff
@@ -380,7 +380,7 @@ ICALL(CULINF_9, "internal_get_cultures", ves_icall_System_Globalization_CultureI
 ICALL_TYPE(REGINF, "System.Globalization.RegionInfo", REGINF_2)
 ICALL(REGINF_2, "construct_internal_region_from_name", ves_icall_System_Globalization_RegionInfo_construct_internal_region_from_name)
 
-#if defined(ENABLE_MONODROID) || defined(ENABLE_MONOTOUCH)
+#if defined(ENABLE_MONODROID) || defined(ENABLE_MONOTOUCH) || defined(HOST_WASM)
 ICALL_TYPE(DEFLATESTREAM, "System.IO.Compression.DeflateStreamNative", DEFLATESTREAM_1)
```
Comment (vargaz) [inline mono/metadata/icall-def.h]: Ditto.

File: mono/metadata/icall.c [hunk 4]
```diff
@@ -7653,7 +7653,7 @@ ves_icall_System_IO_get_temp_path (MonoError *error)
 	return mono_string_new_handle (mono_domain_get (), g_get_tmp_dir (), error);
 }
 
-#if defined(ENABLE_MONODROID) || defined(ENABLE_MONOTOUCH)
+#if defined(ENABLE_MONODROID) || defined(ENABLE_MONOTOUCH) || defined(HOST_WASM)
```
Comment (vargaz) [inline mono/metadata/icall.c]: Ditto.

File: mono/metadata/icall-decl.h (hunk 2 above)
Comment (kjpou1) [reply mono/metadata/icall-decl.h]: Addressed in commit.

File: mono/metadata/icall-def.h (hunk 3 above)
Comment (kjpou1) [reply mono/metadata/icall-def.h]: Addressed in commit.

File: mono/metadata/icall.c (hunk 4 above)
Comment (kjpou1) [reply mono/metadata/icall.c]: Addressed in commit.

**Comment by kjpou1** [comment]
@monojenkins build Linux WebAssembly

**Comment by kjpou1** [comment]
@monojenkins build failed

**Comment by kjpou1** [comment]
@monojenkins build Linux WebAssembly

**Comment by kjpou1** [comment]
@monojenkins build Linux WebAssembly

**Comment by kjpou1** [comment]
@monojenkins build Linux WebAssembly

**Comment by kjpou1** [comment]
@monojenkins build Linux WebAssembly

**Comment by kjpou1** [comment]
> Looking great. But before we land this let's get some real numbers on the release build size increase and see if we can make this a build time option.

Updated in PR comments with sizes for both debug and release builds.

**Comment by kjpou1** [comment]
@monojenkins build failed

**Comment by kjpou1** [comment]
> see if we can make this a build time option.

committed a mod to packager to enable zlib support
   -   --enable-zlib (enable the use of zlib for System.IO.Compression support)
           type: bool  default: false

File: sdks/wasm/docs/packager.md [hunk 5]
```diff
@@ -51,6 +52,8 @@ Additional options (--option/--no-option):
         type: bool  default: true
   --linker-verbose (set verbose option on linker)
         type: bool  default: false
+  --enable-zlib (enable the use of zlib for System.IO.Compression support)
+        type: bool  default: false
```
Comment (vargaz) [inline sdks/wasm/docs/packager.md]: Perhaps this should just be '--zlib'.

Comment (kjpou1) [reply sdks/wasm/docs/packager.md]: Addressed in recent commit

**Comment by kjpou1** [comment]
@monojenkins build failed

**Comment by kjpou1** [comment]
@monojenkins build failed

**Comment by kjpou1** [comment]
@monojenkins build failed

**Pull request merged**